packages/flutter/lib/src/widgets/binding.dart

```diff
@@ -753,18 +748,23 @@ mixin WidgetsBinding on BindingBase, ServicesBinding, SchedulerBinding, GestureB
       return true;
     }());
 
-    if (_needToReportFirstFrame && _reportFirstFrame) {
+    TimingsCallback firstFrameCallback;
+    if (_needToReportFirstFrame) {
       assert(!_firstFrameCompleter.isCompleted);
 
-      TimingsCallback firstFrameCallback;
       firstFrameCallback = (List<FrameTiming> timings) {
+        assert(sendFramesToEngine);
         if (!kReleaseMode) {
           developer.Timeline.instantSync('Rasterized first useful frame');
           developer.postEvent('Flutter.FirstFrame', <String, dynamic>{});
         }
         SchedulerBinding.instance.removeTimingsCallback(firstFrameCallback);
+        firstFrameCallback = null;
         _firstFrameCompleter.complete();
       };
+      // Callback is only invoked when [Window.render] is called. When
+      // [sendFramesToEngine] is set to false during the frame, it will not
+      // be called and we need to remove the callback (see below).
       SchedulerBinding.instance.addTimingsCallback(firstFrameCallback);
```

Comment:
weird that we use addTimingsCallback for this, it seems like that would introduce a lot of latency...

Reply:
The engine fires the first timings callback as soon as possible without batching or adding any other unnecessary latency. The remaining latency is probably unavoidable: only the C++ code in the GPU thread knows when the first frame is rasterized, and it has to notify the UI thread to trigger a Dart function callback. I think that it's more important to make sure that the first frame callback is fired correctly after the rasterization, than to get an early callback where the first frame is not rasterized yet?